File: store-admin/src/main/java/org/delivery/storeadmin/config/rabbitmq/RabbitMqConfig.java
package org.delivery.storeadmin.config.rabbitmq;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.core.AcknowledgeMode;
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.amqp.support.converter.MessageConverter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import org.springframework.amqp.core.Queue;

@Configuration
public class RabbitMqConfig {

    // Queue 설정 (durable = true)
    @Bean
    public Queue deliveryQueue() {
        return new Queue("delivery.queue", true); // durable=true → 서버 재시작 시에도 유지
    }

    // JSON 직렬화 메시지 컨버터
    @Bean
    public MessageConverter messageConverter(ObjectMapper objectMapper){
        return new Jackson2JsonMessageConverter(objectMapper);
    }

    // RabbitTemplate도 필요하면 여기에 추가 가능 (발송용)
    //수동 ACK을 위한 리스너 컨테이너 팩토리
        @Bean
        public SimpleRabbitListenerContainerFactory rabbitListenerContainerFactory(
            ConnectionFactory connectionFactory,
            MessageConverter messageConverter
        ) {
            SimpleRabbitListenerContainerFactory factory = new SimpleRabbitListenerContainerFactory();
            factory.setConnectionFactory(connectionFactory);
            factory.setMessageConverter(messageConverter);
            factory.setAcknowledgeMode(AcknowledgeMode.MANUAL); // 수동 ack 설정
            return factory;
        }
}

File: api/src/main/kotlin/org/delivery/api/config/rabbitmq/RabbitMqConfig.kt
package org.delivery.api.config.rabbitmq

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.amqp.core.Binding
import org.springframework.amqp.core.BindingBuilder
import org.springframework.amqp.core.DirectExchange
import org.springframework.amqp.core.Queue
import org.springframework.amqp.rabbit.config.SimpleRabbitListenerContainerFactory
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.amqp.core.AcknowledgeMode

@Configuration
class RabbitMqConfig(@Qualifier("objectMapper") private val objectMapper: ObjectMapper) {

    @Bean
    fun directExchange(): DirectExchange{
        return DirectExchange("delivery.exchange")
    }

    @Bean
    fun queue(): Queue {
        return Queue("delivery.queue")
    }

    @Bean
    fun binding(
        directExchange: DirectExchange,
        queue: Queue
    ): Binding {
        return BindingBuilder
            .bind(queue)
            .to(directExchange)
            .with("delivery.key")

    }

    // end queue 설정

    @Bean
    fun rabbitTemplate(
        connectionFactory: ConnectionFactory,
        messageconverter: MessageConverter
    ) : RabbitTemplate{
        val rabbitTemplate : RabbitTemplate = RabbitTemplate().apply {
            setConnectionFactory(connectionFactory)
            setMessageConverter(messageconverter)
        }

        return rabbitTemplate
    }

    @Bean
    fun messageConverter(
        objectMapper: ObjectMapper
    ): MessageConverter{
        return Jackson2JsonMessageConverter(objectMapper)
    }

    //수동 ack 설정 포함한 리스너 팩토리 추가
        @Bean
        fun rabbitListenerContainerFactory(
            connectionFactory: ConnectionFactory
        ): SimpleRabbitListenerContainerFactory {
            return SimpleRabbitListenerContainerFactory().apply {
                setConnectionFactory(connectionFactory)
                setAcknowledgeMode(AcknowledgeMode.MANUAL) // 수동 ack
            }
        }
}